pkg/initializer/util/csrUtil.go

```diff
@@ -0,0 +1,163 @@
+		serviceName,
+		fmt.Sprintf("%s.%s", serviceName, namespace),
+		fmt.Sprintf("%s.%s.svc", serviceName, namespace),
+	}
+	requestPem, err := certUtil.GenerateCSRPem(key, serviceName, dnsNames)
+	if err != nil {
+		return nil, err
+	}
+
+	csr := generateCSR(serviceName, namespace, ownerRefs, requestPem)
+	createdCSR, err := kubeCli.CertificatesV1beta1().CertificateSigningRequests().Create(csr)
+	if err != nil {
+		return nil, err
+	}
+	klog.Infof("success to create csr[%s]", csrName)
+
+	_, err = approveCSR(createdCSR, kubeCli, timeout)
+	if err != nil {
+		return nil, err
+	}
+	err = waitUntilCsrApproved(csrName, kubeCli, timeout)
+	if err != nil {
+		return nil, err
+	}
+	klog.Infof("success to approve csr[%s]", csrName)
+
+	approvedCSR, err := kubeCli.CertificatesV1beta1().CertificateSigningRequests().Get(csrName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	return approvedCSR, nil
+}
+
+func approveCSR(csr *certificates.CertificateSigningRequest, kubeCli kubernetes.Interface, timeout time.Duration) (*certificates.CertificateSigningRequest, error) {
+	err := waitUntilCsrExist(csr.Name, kubeCli, timeout)
+	if err != nil {
+		return nil, err
+	}
+	csr = generateApproveCSR(csr)
+	approvedCSR, err := kubeCli.CertificatesV1beta1().CertificateSigningRequests().UpdateApproval(csr)
+	if err != nil {
+		return nil, err
+	}
+	return approvedCSR, nil
+}
+
+func generateCSR(service, namespace string, ownerRefs []metav1.OwnerReference, requestPem []byte) *certificates.CertificateSigningRequest {
+	csr := &certificates.CertificateSigningRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            fmt.Sprintf("%s.%s", service, namespace),
+			OwnerReferences: ownerRefs,
+		},
+		Spec: certificates.CertificateSigningRequestSpec{
+			Request: requestPem,
+			Usages: []certificates.KeyUsage{
+				certificates.UsageDigitalSignature,
+				certificates.UsageServerAuth,
+				certificates.UsageKeyEncipherment,
+			},
+			Groups: []string{
+				"system:authenticated",
+			},
+		},
+	}
+	return csr
+}
+
+func waitUntilCsrDelete(csrName string, kubeCli kubernetes.Interface, timeout time.Duration) error {
+	err := kubeCli.CertificatesV1beta1().CertificateSigningRequests().Delete(csrName, &metav1.DeleteOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	return wait.PollImmediate(1*time.Second, timeout, func() (done bool, err error) {
+		_, err = kubeCli.CertificatesV1beta1().CertificateSigningRequests().Get(csrName, metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return true, nil
+			}
+			return false, err
+		}
+		return false, nil
+	})
+}
+
+func waitUntilCsrExist(csrName string, kubeCli kubernetes.Interface, timeout time.Duration) error {
+
+	return wait.PollImmediate(1*time.Second, timeout, func() (done bool, err error) {
+		_, err = kubeCli.CertificatesV1beta1().CertificateSigningRequests().Get(csrName, metav1.GetOptions{})
+		if err != nil {
+			return false, nil
```

Comment:
errors other than `notFound` should not be ignored

Reply:
> errors other than `notFound` should not be ignored

In this case, I think there is no difference between `notFound` error and other error.

Reply:
`notFound` is one of the expected situations, other unexpected errors should either be logged or returned, or you will end up with getting the timeout message and have no idea about what really happens.

Reply:
fix